### ISSUE TITLE
Illegal action handling in policy gradient algorithms (QPG, RM, RPG and A2C)

### ISSUE THREAD

**Issue by perpetualglow** [deepmind/open_spiel#422]
Hi, there are TODO comments in the policy_gradient.py file who indicate that legal actions still need to be accounted for when computing the losses to update both the critic as well as the actor network.
I am not quite sure if this is even necessary, and if it is, how to do it myself. Would it work if I left the critic update as is, and for the actor update added masked legal actions to all the logits (-infinity when invalid, 0 when valid)?

Any info is much appreciated :) 
Edit: adding the mask to the logits seems to work for RPG at least, learning slowly but steady.

**Comment by lanctot** [comment]
Firstly: thanks for looking into this.. I think we should finally fix it.. this is long overdue! 

I spoke to @fzvinicius about this a few months back. IIRC there was nothing to do on A2C, but the others indeed need to take into account illegal actions for the policy. 

> I am not quite sure if this is even necessary, and if it is, how to do it myself. Would it work if I left the critic update as is, and for the actor update added masked legal actions to all the logits (-infinity when invalid, 0 when valid)?

I believe so. Vinicius, can you confirm?

**Comment by perpetualglow** [comment]
Cheers @lanctot, I've been reading your papers!
Have you tested any of the policy gradient algorithms implemented here for cooperative or mixed cooperative/competitive environments? It was mentioned that the similarity to CFR could help in deriving convergence guarantees for cooperative games, but I guess it's unlikely that these vanilla implementations will actually converge when cooperation is needed. 
I want to try them out on a 4-player trick taking card game (2 teams), since my PPO implementation doesn't do too well once I introduce coop.

**Comment by elkhrt** [comment]
You might be interested in tinybridge4p which is a very small version of
the bidding phase of contract bridge (two teams of two).
I haven't done any real research using it, but the uncontested (2p) version
is non-trivial, and I believe the 4p version has some interesting
additional features, including some positions where mixed strategies are
optimal.
There's a playthrough here:
https://github.com/deepmind/open_spiel/blob/master/open_spiel/integration_tests/playthroughs/tiny_bridge_4p.txt

**Comment by lanctot** [comment]
Hi @perpetualglow, no we have not.. I'b be curious how well they'd do there. It took quite a lot of computation and other things (like population based training, recurrent nets, IMPALA) to get policy gradients to work in Hanabi, but that is quite a large game. 

I'd be delighted if you can submit these fixes as a PR.

BTW you might be interested in this paper: https://arxiv.org/abs/2008.06495

**Comment by perpetualglow** [comment]
Thanks, very cool paper!!

I will revisit this soon and issue a PR, I'm not a particularly good programmer though and have no experience with Tensorflow apart from this, but I think my 'solution' that I mentioned earlier, is relatively clean. RPG and QPG haven't performed particularly well though, only slightly edging out PPO after millions of iterations for very small games, and still quite a ways away from a Nash equilibrium. Performance doesn't seem to scale very well to larger games either, but I have noticed that it's quite sensitive to parameter choices. SGD also seems better (but infinitely slower) than an Adam optimizer, the momentum really screws training over sometimes.

Since my agents are symmetric I had the idea to train with self play by only using one agent, giving it the data of all players and hoping it improves, otherwise I'm worried the different networks diverge from each other due to training randomness. Initially this seemed to be better but I'm having doubts, when using adam optimization it will actually start playing worse after reaching a local minimum. But I need to spend some more time on the whole thing.
Note that all of this is even on a modified game with 2 players, if I don't get anywhere, I likely won't try cooperation.

If anyone reading this is trying to design a bot for trick taking card games, MCTS approaches are easily the way to go. I tried NFSP too, but scaling that to full games is close to impossible I think. I am tempted to fool around with your Deep CFR implementation but I can already feel that it's gonna be extremely sensitive to parameters with ultimately disappointing results.

Sorry for going off topic, once I get back on this, I will make my code PR friendly!

**Comment by lanctot** [comment]
Hi @perpetualglow great to hear from you again. Indeed scaling all of the algorithms to large games is non-trivial, we can certainly relate to the challenges there :)

Since our last communication @elkhrt and colleagues have written a paper about bridge bidding that might interest you: https://arxiv.org/abs/2011.14124

I will follow-up by tagging the people I know working on RL for trick taking card games and scaling.

**Comment by lanctot** [comment]
I know of a few people working on RL for trick-taking games, I will tag them now as they might have some advice or be otherwise interested.

@elkhrt has been working on bridge. @inejc has been working on Slovenian Tarok. @solinas and @flostim have been working on Skat.

In the summer, @solinas and @jhtschultz led a project to get NFSP running on hearts. It was a lot of fun. It became clear to us that some ground work is necessary to get this to work at scale. Most recently @jhtschultz has been working on using Acme to get a large-scale NFSP implementation (using OpenSpiel games), which is super promising. It is not quite ready yet but quite close.. and the integration with Acme already done which is the first step:
